### PR TITLE
fix: take SRCNN to the reference degradation order too

### LINKED ISSUE #245: SRCNN degrades an extracted patch; the authors degrade the image and then extract
## The defect

SRCNN's training path degrades an **extracted HR patch**. The authors degrade the **whole
image** and extract patches from the result.

Ours, in `datasets/srcnn.py::TrainDataset.__getitem__`:

```python
hr_subimg  = <33x33 slice of the HR image>
lr_subimg  = _degrade(hr_subimg, self.scale)     # degrade the PATCH
```

Theirs, in the released `demo_SR.m` / `generate_train.m` order:

```matlab
im_gnd = modcrop(im, up_scale);
im_l   = imresize(im_gnd, 1/up_scale, 'bicubic');   % degrade the IMAGE
im_b   = imresize(im_l,   up_scale,   'bicubic');
% ... only then are 33x33 sub-images taken from im_gnd / im_b
```

These are not the same operation. Bicubic interpolation's kernel support extends past the
patch boundary, so degrading a 33x33 crop resolves those taps against the crop's edge
handling instead of against the real neighbouring pixels.

## Measured

Smooth, image-like content (noise exaggerates this heavily — on random input the same
comparison reads max 36–59 levels, which is not representative):

| scale | whole patch max\|diff\| | mean | 4 px border mean | interior mean |
|---:|---:|---:|---:|---:|
| 2 | 3 | 0.519 | 0.663 | 0.424 |
| **3** | **5** | 0.333 | **0.831** | **0.000** |
| 4 | 8 | 0.904 | 1.574 | 0.450 |

**At scale 3 the interior is bit-identical and the entire difference is the border.** That is
the signature of the kernel reaching outside the patch and nothing else. At scales 2 and 4 the
interior differs too, because `sub_img_size = 33` is not a multiple of those scales, so the LR
sampling grid also shifts.

A 4 px border is **43%** of a 33x33 patch's pixels.

## Why it is worth filing rather than shrugging at

The differences are small in absolute terms, and this is deliberately filed as a fidelity gap
rather than a correctness bug. But this is a paper-reproduction framework, and the standing
rule is that a figure is comparable only to one computed the same way. Here the *training
inputs* are not produced the way the authors produced theirs — so a reproduction gap has a
candidate explanation that is not the model.

It also means `sub_img_size` and `scale` are silently coupled: `33 % 3 == 0` makes scale 3
much better behaved than 2 or 4, and nothing anywhere says so.

## Relationship to the modcrop issue

A separate issue covers SRCNN *validation* skipping modcrop. That one is about **which pixels
exist**; this one is about **when the degradation is applied relative to patching**. Fixing
the first does not fix this — after it, train and eval agree on the modcrop convention and
still disagree on this axis.

## Scope note

Not free. Degrading whole images at load time, or caching an LR plane alongside the HR one,
changes the data path this project measured at 98.94% decode-dominated — so the fix needs a
throughput measurement, not just a correctness one.

## Acceptance

- [ ] SRCNN training patches are cut from a degraded whole image, matching the released order
- [ ] A test on a patch whose border demonstrably differs today, proven RED first
- [ ] `sub_img_size` / `scale` coupling is either removed or documented where it is chosen
- [ ] The data path's throughput is measured before and after on this machine

### PR DESCRIPTION
**Pull request by YeapJieShen**
Closes #245. Stacked on #262.

Takes SRCNN to the same reference degradation order #260 gave SRResNet, reverting
the hold I placed on it — and documents what that hold was actually about.

## Why the hold is lifted

It was never a correctness objection. The order is right: the authors' released
`demo_SR.m` is explicit (`modcrop` -> `imresize` down -> `imresize` up, *then*
extraction), and degrading an extracted patch resolves the bicubic kernel's edge
taps against the patch border. At x3 the patch interior is bit-identical and the
**entire** difference is a 4px border — 43% of a 33x33 patch.

The objection was **memory**, and memory is a property of the box, not of the
code. SRCNN is pre-upsampled, so its degraded plane is HR-sized rather than
`1/scale**2` of it, and caching it roughly doubles the working set a shuffling
loader touches. Measured on a 12 GB box against a 12.6 GB working set:

| | before | after |
|---|---:|---:|
| real training loop, 400 steps | 21.60 it/s | **1.02 it/s** |

GPU at 13%, 36% iowait, 992 MB/s sustained read. That is a **step change at the
point the working set stops fitting in page cache**, not a gradient — so it is a
sizing requirement, not a reason to ship the wrong degradation.

## 🚨 Read before merging: this needs RAM

**Budget roughly two bytes of RAM per HR pixel in the training set, plus
headroom.** For DIV2K-800 that is ~12.6 GB, so **16 GB is the practical floor and
12 GB is not enough.** On a box below it, SRCNN training will be I/O-bound and
roughly 20x slower, silently — there is no error, only a slow run.

Both `datasets/srcnn.py` and `datasets/derived_cache.py` now state this where the
choice is made, and name the symptom (low GPU utilisation, high iowait, sustained
disk read) because **it cannot be seen in a per-item benchmark** — per-item
*work* is not what changes.

I did not add a runtime memory check. A host-specific guard shipping in a public
framework is something this project has declined before, and the honest version
of it would have to read total RAM and estimate a working set on every platform.
Say the word if you would rather have an opportunistic warning that degrades to
silence where it cannot measure.

## What ships

- SRCNN slices both sub-images out of whole-image planes at the same grid
  position; no kernel reaches past a patch edge.
- `derived_kind` is required again on the shared base — the `None` opt-out had no
  remaining caller, and speculative generality is what the deletion test rejects.
- The `sub_img_size` / `scale` coupling **dissolves**: with no per-patch sampling
  grid left to shift, `33 % scale` stops mattering. There is a test for it across
  x2 / x3 / x4.

## Tests

The whole-image-order tests come back, including the one that ships its own
mutation — it recomputes the *old* per-patch order and fails if the dataset
returns it, so it cannot pass on either implementation.

Full suite: **1,084 passed, 2 skipped** (both Windows-only liveness paths).

### CODE REVIEW

**Review by YeapJieShen — COMMENTED** [review]
Human reviewed, all good.
